haskell-src/Concordium/Types/Execution.hs

```diff
@@ -1894,18 +1895,39 @@ data TransactionSummary' a = TransactionSummary
       tsResult :: !a,
       tsIndex :: !TransactionIndex
     }
-    deriving (Eq, Show, Generic)
+    deriving (Eq, Show, Generic, Functor)
 
 -- | A transaction summary parameterized with an outcome of a valid transaction
 --  containing either a 'TxSuccess' or 'TxReject'.
 type TransactionSummary = TransactionSummary' ValidResult
 
+-- | An abstraction for constructing the result of a transaction.
+--  This is instantiated by both 'ValidResult' and 'ValidResultWithReturn'.
+class TransactionResult a where
+    -- | Construct a successful result.
+    transactionSuccess :: [Event] -> a
+
+    -- | Construct a rejecting result.
+    transactionReject :: RejectReason -> a
+
+    -- | Attach a return value to the transaction.
+    setTransactionReturnValue :: BS.ByteString -> a -> a
+
+    -- | Construct a result from a 'ValidResult'.
+    fromValidResult :: ValidResult -> a
```

Comment:
`setTransactionReturnValue` seems like an error-prone operation. Especially since this is implemented for ValidResult where that operation does nothing.

Similarly it seems error-prone to have `fromValidResult` since it defaults a value which is usually a source of problems.

Maybe there's no better solution, but this is not a nice abstraction.